File: src/test/java/me/saro/jwt/java/alg/PsThreadTest.java
package me.saro.jwt.java.alg;


import me.saro.jwt.alg.ps.JwtPsAlgorithm;
import me.saro.jwt.core.Jwt;
import me.saro.jwt.core.JwtAlgorithm;
import me.saro.jwt.core.JwtKey;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.time.OffsetDateTime;
import java.util.*;

@DisplayName("[Java] PS Thread And Random KID Test")
public class PsThreadTest {

    public int randomKeyBit() {
        return List.of(2048, 3072, 4096).get((int)(Math.random() * 3));
    }

    @Test
    @DisplayName("Thread And Random KID Test")
    public void t1() {
        List<JwtPsAlgorithm> algs = List.of(Jwt.PS256, Jwt.PS384, Jwt.PS512);
        Map<String, JwtAlgorithm> algMap = algs.stream().collect(HashMap::new, (m, a) -> m.put(a.getFullname(), a), HashMap::putAll);
        HashMap<String, JwtKey> keys = new HashMap<String, JwtKey>();
        ArrayList<String> jwts = new ArrayList<String>();

        for (int i = 0 ; i < 30 ; i++) {
            JwtPsAlgorithm alg = algs.get((int)(Math.random() * 3));
            String kid = UUID.randomUUID().toString();
            JwtKey key = alg.newRandomJwtKey( randomKeyBit());
            keys.put(kid, key);
            jwts.add(Assertions.assertDoesNotThrow(() -> Jwt.builder().kid(kid).id("abc").expire(OffsetDateTime.now().plusMinutes(30)).toJwt(alg, key)));
        }

        jwts.parallelStream().map(jwt -> Assertions.assertDoesNotThrow(() ->
                Jwt.parse(jwt, node -> {
                    JwtAlgorithm alg = algMap.get(node.getAlgorithm());
                    JwtKey key = keys.get(node.getKid());
                    return alg.with(key);
                })
        )).forEach(JwtNode -> Assertions.assertEquals("abc", JwtNode.getId()));
        System.out.println("done");
    }
}

File: src/test/kotlin/me/saro/jwt/kotlin/alg/PsThreadTest.kt
package me.saro.jwt.kotlin.alg

import me.saro.jwt.alg.ps.JwtPs
import me.saro.jwt.core.Jwt
import me.saro.jwt.core.JwtClaims
import me.saro.jwt.core.JwtClaims.Companion.create
import me.saro.jwt.core.JwtKey
import me.saro.jwt.exception.JwtException
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.time.OffsetDateTime
import java.util.*

@DisplayName("[Kotlin] PS Thread And Random KID Test")
class PsThreadTest {

    fun randomKeyBit() = listOf(2048, 3072, 4096)[(Math.random() * 3).toInt()]

    @Test
    @DisplayName("Thread And Random KID Test")
    fun t1() {
        val algs = listOf(Jwt.ps256(), Jwt.ps384(), Jwt.ps512())
        val keys = HashMap<String?, JwtKey>()
        val jwts = ArrayList<String>()
        for (i in 0..29) {
            val alg = algs[(Math.random() * 3).toInt()]
            val kid = UUID.randomUUID().toString()
            val key = alg.newRandomJwtKey(randomKeyBit())
            keys[kid] = key
            val jc = create()
            jc.id("abc")
            jc.expire(OffsetDateTime.now().plusMinutes(30))
            jwts.add(Assertions.assertDoesNotThrow<String> { alg.toJwt(key, jc, kid) })
        }
        jwts.parallelStream().forEach { jwt: String? ->
            // use alg.toJwtHeader
            // but this case is unknown alg
            // use JwtUtils.toJwtHeader
            val jh = Jwt.toJwtHeader(jwt)
            var _alg: JwtPs? = null
            when (jh.algorithm) {
                "PS256" -> _alg = Jwt.ps256()
                "PS384" -> _alg = Jwt.ps384()
                "PS512" -> _alg = Jwt.ps512()
            }
            val alg = _alg
            Assertions.assertNotNull(alg)
            val key = keys[jh.kid]
            Assertions.assertNotNull(key)
            val jc = Assertions.assertDoesNotThrow<JwtClaims> {
                alg!!.toJwtClaims(
                    jwt!!, key
                )
            }
            Assertions.assertThrows(JwtException::class.java) {
                alg!!.toJwtClaims(
                    jwt!!, alg.newRandomJwtKey(randomKeyBit())
                )
            }
            Assertions.assertEquals(jc.id, "abc")
        }
        println("done")
    }
}